app/scripts/modules/azure/securityGroup/clone/cloneSecurityGroup.controller.js

```diff
@@ -6,68 +6,154 @@ module.exports = angular
+      securityGroup.name = name;
+      $scope.namePreview = name;
+    };
+
+    $scope.securityGroup = securityGroup;
+
+    $scope.state = {
+      refreshingSecurityGroups: false,
+    };
 
+    $scope.taskMonitor = taskMonitorService.buildTaskMonitor({
+      application: application,
+      title: 'Updating your security group',
+      modalInstance: $modalInstance,
+      onTaskComplete: onTaskComplete,
+    });
 
     accountService.listAccounts('azure').then(function(accounts) {
       $scope.accounts = accounts;
-      vm.accountUpdated();
+      ctrl.accountUpdated();
     });
 
-    securityGroup.securityGroupIngress = _(securityGroup.inboundRules)
-      .filter(function(rule) {
-        return rule.securityGroup;
-      }).map(function(rule) {
-        return rule.portRanges.map(function(portRange) {
-          return {
-            name: rule.securityGroup.name,
-            type: rule.protocol,
-            startPort: portRange.startPort,
-            endPort: portRange.endPort
-          };
-        });
-      })
-      .flatten()
-      .value();
-
-    securityGroup.ipIngress = _(securityGroup.inboundRules)
-      .filter(function(rule) {
-        return rule.range;
-      }).map(function(rule) {
-        return rule.portRanges.map(function(portRange) {
-          return {
-            cidr: rule.range.ip + rule.range.cidr,
-            type: rule.protocol,
-            startPort: portRange.startPort,
-            endPort: portRange.endPort
-          };
-        });
-      })
-      .flatten()
-      .value();
-
-
-    vm.upsert = function () {
-      vm.mixinUpsert('Clone');
+    ctrl.addRule = function(ruleset) {
+      ruleset.push({
+        name: $scope.securityGroup.name + '-Rule' + ruleset.length,
+        priority: ruleset.length === 0 ? 100 : 100 * (ruleset.length + 1),
+        protocol: 'tcp',
+        access: 'Allow',
+        direction: 'InBound',
+        sourceAddressPrefix: '*',
+        sourcePortRange: '*',
+        destinationAddressPrefix: '*',
+        destinationPortRange: '7001-7001',
+        startPort: 7001,
+        endPort: 7001
+      });
     };
 
-    vm.initializeSecurityGroups();
+    function onApplicationRefresh() {
+      // If the user has already closed the modal, do not navigate to the new details view
+      if ($scope.$$destroyed) {
+        return;
+      }
+      $modalInstance.close();
+      var newStateParams = {
+        name: $scope.securityGroup.name,
+        accountId: $scope.securityGroup.credentials || $scope.securityGroup.accountName,
+        region: $scope.securityGroup.region,
+        provider: 'azure',
+      };
+      if (!$state.includes('**.securityGroupDetails')) {
+        $state.go('.securityGroupDetails', newStateParams);
+      } else {
+        $state.go('^.securityGroupDetails', newStateParams);
+      }
+    }
+
+    function onTaskComplete() {
+      application.refreshImmediately();
+      application.registerOneTimeRefreshHandler(onApplicationRefresh);
```

Comment:
These two calls should probably be replaced with:
```javascript
application.securityGroups.refresh();
```
(sorry, that change just went in this morning)